jgrapht-core/src/main/java/org/jgrapht/graph/specifics/FastLookupDirectedSpecifics.java

```diff
@@ -0,0 +1,131 @@
+/* ==========================================
+ * JGraphT : a free Java graph-theory library
+ * ==========================================
+ *
+ * Project Info:  http://jgrapht.sourceforge.net/
+ * Project Creator:  Barak Naveh (http://sourceforge.net/users/barak_naveh)
+ *
+ * (C) Copyright 2003-2008, by Barak Naveh and Contributors.
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+/* -----------------
+ * FastLookupDirectedSpecifics.java
+ * -----------------
+ * (C) Copyright 2015-2015, by Joris Kinable and Contributors.
+ *
+ * Original Author:  Joris Kinable
+ * Contributor(s):
+ *
+ * $Id$
+ *
+ * Changes
+ * -------
+ */
+package org.jgrapht.graph.specifics;
+
+import org.jgrapht.Graph;
+import org.jgrapht.graph.AbstractBaseGraph;
+import org.jgrapht.util.ArrayUnenforcedSet;
+import org.jgrapht.util.VertexPair;
+
+import java.io.Serializable;
+import java.util.*;
+
+/**
+ * Fast implementation of DirectedSpecifics. This class uses additional data structures to improve the performance of methods which depend
+ * on edge retrievals, e.g. getEdge(V u, V v), containsEdge(V u, V v),addEdge(V u, V v). A disadvantage is an increase in memory consumption.
+ * If memory utilization is an issue, use a {@link DirectedSpecifics} instead.
+ *
+ * @author Joris Kinable
+ */
+public class FastLookupDirectedSpecifics<V,E>
+    extends DirectedSpecifics<V,E>
+    implements Serializable
+{
+    private static final long serialVersionUID = 4089085208843722263L;
+
+    /* Maps a pair of vertices <u,v> to an set of edges {(u,v)}. In case of a multigraph, all edges which touch both u,v are included in the set */
+    protected Map<VertexPair<V>, ArrayUnenforcedSet<E>> touchingVerticesToEdgeMap;
+
+    public FastLookupDirectedSpecifics(AbstractBaseGraph<V, E> abstractBaseGraph)
+    {
+        this(abstractBaseGraph, new LinkedHashMap<>());
+    }
+
+    public FastLookupDirectedSpecifics(AbstractBaseGraph<V, E> abstractBaseGraph, Map<V, DirectedEdgeContainer<V, E>> vertexMap)
+    {
+        super(abstractBaseGraph, vertexMap);
+        this.touchingVerticesToEdgeMap=new HashMap<>();
+    }
+
+
+    /**
+     * @see Graph#getAllEdges(Object, Object)
+     */
+    @Override public Set<E> getAllEdges(V sourceVertex, V targetVertex)
+    {
+        if (abstractBaseGraph.containsVertex(sourceVertex)&& abstractBaseGraph.containsVertex(targetVertex)) {
+            Set<E> edges = touchingVerticesToEdgeMap.get(new VertexPair<>(sourceVertex, targetVertex));
+            return edges == null ? Collections.emptySet() : new ArrayUnenforcedSet<>(edges);
```

Comment:
This is a tricky issue. If we return an immutable set, then the set returned is a *view* of the edges in between (i,j). Imagine that a user invokes deleteAllEdges(getAllEdges(i,j)); this will cause a concurrent modification exception since deleteAllEdges will iterate over the set and invoke deleteEdge(..) on each item in that set. Similarly, we would need to change the implementation of deleteAllEdges(V v1, V v2) for the exact same reason. Nevertheless, I think this would be a good idea though to further enhance performance. What do you think?